include/chainbase/chainbase.hpp

```diff
@@ -533,10 +545,26 @@ namespace chainbase {
             return ret;
          }
 
+         void set_read_only_mode() {
+            _read_only_mode = true;
+         }
+
+         void unset_read_only_mode() {
+            _read_only_mode = false;
+         }
+
       private:
          pinnable_mapped_file                                        _db_file;
          bool                                                        _read_only = false;
 
+         /**
+          * This is a catch-all flag. When set to true, an exception is
+          * throw when modification  attempt is made on chainbase.
+          * This ensures state is not modified by mistake when application
+          * does not intend to change state.
+          */
+         bool                                                        _read_only_mode = false;
```

Comment:
We have both a `_read_only` and a `_read_only_mode`. I'm not sure whether we even use `_read_only` in Leap, but if so, I understand it to be something that is static throughout the lifetime of the database instance, whereas `_read_only_mode` is meant to be dynamic that we can toggle back and forth between rapidly.

It seems like the constructor and the mutators for `_read_only_mode` should respect if the `database` instance is constructed to be read_only and disallow `_read_only_mode` from being false if Chainbase is opened up as read only.

Better yet, if we are not using the read-only mode and the mutex locks (this is something that came from earlier versions of Chainbase used in blockchains prior to EOS) then I would like to see a simplification. Right now we have the `CHAINBASE_REQUIRE_WRITE_LOCK` macro almost every where the check for `_read_only_mode` to throw the exception exists (the only exception is for `set_revision` which should also probably be guarded similarly).

Also, seems like we should protect creating undo sessions and the things that can be done with them.

Reply:
`_read_only` is used in Leap. In `controller::extract_chain_id_from_db`, DB is opened as `chainbase::database::read_only`. I don't see nodeos startup command has DB read only option. I will change to honor `_read_only` when setting `_read_only_mode`.

I could not find `CHAINBASE_CHECK_LOCKING` is enabled in Leap. If it was enabled, I am not sure if the current implementation is correct. I do see `_read_lock_count` or `_write_lock_count` ever is modified. I am also not sure why `&&` is not used in the `if` statements.

```
         void require_read_lock( const char* method, const char* tname )const
         {
            if( BOOST_UNLIKELY( _enable_require_locking & _read_only & (_read_lock_count <= 0) ) )
               require_lock_fail(method, "read", tname);
         }

         void require_write_lock( const char* method, const char* tname )
         {
            if( BOOST_UNLIKELY( _enable_require_locking & (_write_lock_count <= 0) ) )
               require_lock_fail(method, "write", tname);
         }
```

Reply:
I was also struggling with naming `_read_only_mode`, fearing the confusion it might cause to `_read_only`. A different name could be better.

Reply:
Should we remove `CHAINBASE_CHECK_LOCKING` related code from the code base to prevent it from using?

Reply:
@spoonincode What do you think if we remove `CHAINBASE_CHECK_LOCKING`, `CHAINBASE_REQUIRE_READ_LOCK`, and `CHAINBASE_REQUIRE_WRITE_LOCK` related code? The existing code seems incorrect.

Reply:
Thanks. Will do it.